### ISSUE TITLE
Variance check is inconsistent

### ISSUE THREAD

**Issue by hurryabit** [microsoft/pyright#10229]
Pylance version: 2025.2.0

The variance check for the type parameters of generic classes does not take the variance restrictions imposed by instance variables and methods into account properly. For instance, using a type variable `T_co` declared as covariant as an invariant type argument in the type of an instance variable is not flagged.

The following example demonstrates this issue. Pyright accepts this code although it fails at runtime with what basically is a type error there's a `Dog` in a `list[Cat]`:
```python
from abc import ABC
from typing import Generic, TypeVar

T_co = TypeVar("T_co", covariant=True)

class MyList(Generic[T_co]):
    data: list[T_co]

    def __init__(self) -> None:
        self.data = []

class Animal(ABC):
    pass

class Cat(Animal):
    pass

class Dog(Animal):
    pass

def append_dog(animals: MyList[Animal]) -> None:
    animals.data.append(Dog())

cats = MyList[Cat]()
cats.data.append(Cat())

append_dog(cats)
assert all(isinstance(cat, Cat) for cat in cats.data)
```

I'd say this is inconsistent since adding the method
```python
    def append(self, x: T_co) -> None:
        pass
```
to `MyList` gets flagged with an error message reading
> Covariant type variable cannot be used in parameter type

In contrast, the method
```python
    def appender(self) -> Callable[[T_co], None]:
        return lambda x: self.data.append(x)
```
does not get flagged although `T_co` is used in a negative position. In fact, this method allows for adding a `Dog` to a `list[Cat]` again without the type checker picking up on it.

Moreover, the method
```python
    def for_each(self, f: Callable[[T_co], None]) -> None:
        pass
```
is not flagged, which is correct since `T_co` only appears in a positive position. However, the method
```python
    def foo(self, f: Callable[[Callable[[T_co], None]], None]) -> None:
        pass

```
also doesn't get flagged despite `T_co` appearing an a negative position, which is inconsistent with `append` getting flagged.

**Comment by erictraut** [comment]
Pyright is working as designed here, so I don't consider this a bug.

If you manually specify that a type variable is covariant, pyright will treat it as covariant when performing type analysis. As you've pointed out, you can receive incorrect typing results if you specify the wrong variance.

Pyright will validate variance correctness for type variables that parameterize a protocol class, but it doesn't validate variance for standard (nominal) classes. This precedent was established by mypy, and pyright follows suit.

You can see this if you change your definition of `MyList` to a protocol.

```python
class MyList(Protocol[T_co]):
    data: list[T_co]
```

The whole topic of variance is confusing to many Python developers, and it's easy to specify the wrong variance when doing so manually. That's why [PEP 695](https://peps.python.org/pep-0695/) introduced variance inference. I recommend switching to PEP 695 syntax if you're using Python 3.12 or newer.

Here's how that looks.

```python
class MyList[T]:
    data: list[T]

    def __init__(self) -> None:
        self.data = []
```

Now, if you hover over `T` in the class definition, you'll see that pyright correctly infers that it is invariant.

![Image](https://github.com/user-attachments/assets/51d3ac79-3ec5-473d-850a-35d38403fb83)

Or if you still need to support a version of Python older than 3.12, you can use the old syntax but with inferred variance.

```python
from typing_extensions import TypeVar

T = TypeVar("T", infer_variance=True)

class MyList[T]:
    data: list[T]

    def __init__(self) -> None:
        self.data = []
```

**Comment by hurryabit** [comment]
I totally agree with the design decision to accept the declared variance of type parameters for classes without any further checking. Otherwise, there would be no way to override what the type checker infers.

However, pyright is not consistent in this regard. As I pointed out above, the following class definition causes a type error:

```python
T_co = TypeVar("T_co", covariant=True)

class MyList(Generic[T_co]):
    def append(self, x: T_co) -> None:
        pass
```

The use of `T_co` in the signature of the `append` methods gets flagged with

> Covariant type variable cannot be used in parameter type

In spirit of the aforementioned design decision, this should **not** cause a type error.

**Comment by erictraut** [comment]
This error was added a long time ago by mypy as a way to help guide developers to specify variance correctly. As I mentioned above (and in the motivation for PEP 695), many developers find the concept of variance confusing and don't understand it. I added this same message to pyright for the same reason that mypy did.

If this error is generated, it means that you have specified a variance that is incorrect for your usage. If you want to ignore the error and understand the consequences of doing so, you are free to add a `# type: ignore` to suppress the error.
